File: FreePascal/unit.pas
unit &Unit;

interface

type
  PUnit = ^TUnit;
  TUnit = packed record
  public var
    Factor : Double;
    Symbol : string;
  public
    constructor Create( const inBase : TUnit ); overload;
    constructor Create( const inFactor : Double; const inSymbol : string ); overload;
    function Convert( const inValue : Double; const inReference : TUnit ) : Double;
    function Equals( const inRight : TUnit ) : Boolean;
  end;

  POperand = ^TOperand;
  TOperand = packed record
  private var
    m_fValue : Double;
    m_varBase : TUnit;
  private
    function GetValue() : Double; inline;
    function GetBase() : TUnit; inline;
  public
    constructor Create( const inBase : TUnit; const inValue : Double = 0.0 );
    procedure SetValue( const inValue : Double; const inReference : TUnit );
  public
    property Value : Double read GetValue;
    property Base : TUnit read GetBase;
  end;

implementation

{$REGION ' TUnit '}

function TUnit.Equals( const inRight : TUnit ) : Boolean;
begin
  Result := (Factor = inRight.Factor);
end;

constructor TUnit.Create( const inBase : TUnit );
begin
  Factor := inBase.Factor;
  Symbol := inBase.Symbol;
end;

constructor TUnit.Create( const inFactor : Double; const inSymbol : string );
begin
  Factor := inFactor;
  Symbol := inSymbol;
end;

function TUnit.Convert( const inValue : Double; const inReference : TUnit ) : Double;
begin
  Result := (inValue * (Factor / inReference.Factor));
end;

{$ENDREGION}

{$REGION ' TOperand '}

// Private

function TOperand.GetValue() : Double;
begin
  Result := m_fValue;
end;

function TOperand.GetBase() : TUnit;
begin
  Result := m_varBase;
end;

// Public

constructor TOperand.Create( const inBase : TUnit; const inValue : Double );
begin
  m_varBase := inBase;
  m_fValue := inValue;
end;

procedure TOperand.SetValue( const inValue : Double; const inReference : TUnit );
begin
  m_fValue := inReference.Convert( inValue, m_varBase );
end;

{$ENDREGION}

end.

File: C#/unit.cs
namespace metric_cnv
{
   public class Unit
   {

      #region ' Public Methods '

      public Unit()
      {
         Factor = 0.0;
         Symbol = string.Empty;
      }

      public Unit( double inFactor, string inSymbol )
      {
         Factor = inFactor;
         Symbol = inSymbol;
      }

      public Unit( Unit inBase )
      {
         Factor = inBase.Factor;
         Symbol = inBase.Symbol;
      }

      public override bool Equals( object? inObject )
      {
         return inObject as Unit == this;
      }

      public override int GetHashCode()
      {
         return Factor.GetHashCode() | Symbol.GetHashCode();
      }

      public double Convert( double inValue, Unit inReference )
      {
         return inValue * (Factor / inReference.Factor);
      }

      public static bool operator ==( Unit inLeft, Unit inRight )
      {
         return inLeft.Factor == inRight.Factor;
      }

      public static bool operator !=( Unit inLeft, Unit inRight )
      {
         return inLeft.Factor != inRight.Factor;
      }

      #endregion

      #region ' Properties '

      public double Factor { get; }
      public string Symbol { get; }

      #endregion
   }

   public class Operand
   {

      public Operand()
      {
         Base = null;
         Value = 0.0;
      }

      public Operand( Unit inBase, double inValue = 0.0 )
      {
         Base = new Unit( inBase );
         Value = inValue;
      }

      public double Convert( Unit inReference ) => Base.Convert( Value, inReference );

      public static explicit operator double( Operand inOperand )
      {
         return inOperand.Value;
      }

      public static Operand operator <<( Operand inLeft, Operand inRight )
      {
         return new Operand( inLeft.Base, inRight.Convert( inLeft.Base ) );
      }

      public static Operand operator <<( Operand inLeft, double inValue )
      {
         return new Operand( inLeft.Base, inValue );
      }

      public Unit Base { get; }
      public double Value { get; private set; }
   }
}